### PR TITLE
Documentation: Update instructions on deploying to GitHub Pages

### PR DESCRIPTION
**Pull request by eugenepeh**
Ready for review

### CODE REVIEW

**Comment by eugenepeh** [comment]
@damithc as requested in gitter. The 2. mentioned by @pyokagan not included yet as there's the issue of html link in PDF reflect local directory.

**Comment by eugenepeh** [comment]
Oops [2/4] and [4/4] should be together, but its going get squashed into one anyway right?

**Comment by eugenepeh** [comment]
Hi @Zhiyuan-Amos, can help to review this and #676 ? :Smile:

**Comment by Zhiyuan-Amos** [comment]
9:30am; omw to gym now x) I'm not too good with reviewing documentation tho, but will try ><

**Comment by eugenepeh** [comment]
ahh okay, not sure if prof needs it soon, so just help me take a look if the instruction is clear enough when you're free please 😄

File: docs/DeveloperGuide.adoc [hunk 1]
```diff
@@ -420,7 +420,7 @@ Alternatively, you can download the AsciiDoc plugin for IntelliJ, which allows y
 
 === Publishing Documentation
 
-See <<UsingGithubPages#, UsingGithubPages.adoc>> to learn how to use GitHub Pages to publish documentation to the project site.
+See https://pages.github.com[GitHub Pages] to learn how to use GitHub Pages to publish documentation to the project site.
```
Comment (Zhiyuan-Amos) [inline docs/DeveloperGuide.adoc]: `See GitHub Pages to learn how to use GitHub Pages`... sounds weird right :P Perhaps something like `See GitHub Pages to learn how to publish documentation to the project site` will do.

File: docs/DeveloperGuide.adoc [hunk 2]
```diff
@@ -437,8 +437,8 @@ We use https://www.google.com/chrome/browser/desktop/[Google Chrome] for convert
 
 Here are the steps to convert the project documentation files to PDF format.
 
-.  Make sure you have set up GitHub Pages as described in <<UsingGithubPages#setting-up, UsingGithubPages.adoc>>.
-.  Using Chrome, go to the <<UsingGithubPages#viewing-the-project-site, GitHub Pages version>> of the documentation file. e.g. For <<UserGuide#, UserGuide.adoc>>, the URL will be `\https://<your-username-or-organization-name>.github.io/addressbook-level4/docs/UserGuide.html`.
+.  Make sure you have set up GitHub Pages as described in <<DeveloperGuide#54-publishing-documentation, 5.4>>.
+.  Using Chrome, go to the HTML link of the documentation file. e.g. For <<UserGuide#, UserGuide.adoc>>, the URL will be `\https://<your-username-or-organization-name>.github.io/addressbook-level4/docs/UserGuide.html`.
```
Comment (Zhiyuan-Amos) [inline docs/DeveloperGuide.adoc]: `HTML link` sounds weird, cos it refers to the `<a href>` thing. Maybe something like `Using Chrome, go to the URL of the documentation file...`

File: docs/DeveloperGuide.adoc [hunk 3]
```diff
@@ -422,6 +422,10 @@ Alternatively, you can download the AsciiDoc plugin for IntelliJ, which allows y
 
 See https://pages.github.com[GitHub Pages] to learn how to use GitHub Pages to publish documentation to the project site.
 
+.  Convert the AsciiDoc files in `docs` to HTML format, see <<UsingGradle#rendering-asciidoc-files, UsingGradle.adoc>>.
```
Comment (Zhiyuan-Amos) [inline docs/DeveloperGuide.adoc]: Can rephrase to something like `See UsingGradle.adoc to learn how to convert...`

File: docs/DeveloperGuide.adoc [hunk 4]
```diff
@@ -422,6 +422,10 @@ Alternatively, you can download the AsciiDoc plugin for IntelliJ, which allows y
...
+.  Copy all the generated files and folders (E.g. HTML files and images folder) in `build/docs` and place it into the `project folder` that you have setup for GitHub Pages.
```
Comment (Zhiyuan-Amos) [inline docs/DeveloperGuide.adoc]: Not sure whether `(E.g. HTML files and images folder)` is necessary since you've already mentioned to copy `all the generated files and folders in build/docs`

File: docs/DeveloperGuide.adoc [hunk 5]
```diff
@@ -422,6 +422,10 @@ Alternatively, you can download the AsciiDoc plugin for IntelliJ, which allows y
...
+.  Push it and you will be able to see your documentation on `\https://<your-username-or-organization-name>.github.io`
```
Comment (Zhiyuan-Amos) [inline docs/DeveloperGuide.adoc]: Missing period.

Comment (Zhiyuan-Amos) [reply docs/DeveloperGuide.adoc]: `Push it` sounds weird haha, maybe something like `Push the changes and you will see your documentation published on...`

File: docs/DeveloperGuide.adoc [hunk 6]
```diff
@@ -437,8 +441,8 @@ We use https://www.google.com/chrome/browser/desktop/[Google Chrome] for convert
 
 Here are the steps to convert the project documentation files to PDF format.
 
-.  Make sure you have set up GitHub Pages as described in <<UsingGithubPages#setting-up, UsingGithubPages.adoc>>.
-.  Using Chrome, go to the <<UsingGithubPages#viewing-the-project-site, GitHub Pages version>> of the documentation file. e.g. For <<UserGuide#, UserGuide.adoc>>, the URL will be `\https://<your-username-or-organization-name>.github.io/addressbook-level4/docs/UserGuide.html`.
+.  Make sure you have set up GitHub Pages as described in <<DeveloperGuide#54-publishing-documentation, 5.4>>.
```
Comment (Zhiyuan-Amos) [inline docs/DeveloperGuide.adoc]: Over here I'm not sure whether we should use `5.4` or `Publishing Documentation`. That's cos on line 363, there's a link to `Configuration` in which the actual word "Configuration" is used, instead of the title number.

File: docs/DeveloperGuide.adoc [hunk 7]
```diff
@@ -420,7 +420,11 @@ Alternatively, you can download the AsciiDoc plugin for IntelliJ, which allows y
 
 === Publishing Documentation
 
-See <<UsingGithubPages#, UsingGithubPages.adoc>> to learn how to use GitHub Pages to publish documentation to the project site.
+See https://pages.github.com[GitHub Pages] to learn how to publish documentation to the project site.
+
+.  See <<UsingGradle#rendering-asciidoc-files, UsingGradle.adoc>> to learn how to convert the AsciiDoc files in `docs` to HTML format.
+.  Copy all the generated files and folders in `build/docs` and place it into the `project folder` that you have setup for GitHub Pages.
+.  Push the changes and you will see your documentation published on `\https://<your-username-or-organization-name>.github.io`.
```
Comment (damithc) [inline docs/DeveloperGuide.adoc]: The alternative:
>One time setup
>1. Pull gh-pages from upstream repo
>2. Push to your fork 
>3. Go to github settings and choose gh-pages option. The rendered pages are available for viewing now.
>3. Set up personal token -> future changes will be automatically available for viewing

@pyokagan what do you think? I think the set up cost is worth the future convenience. The current way of manually pushing to gh-pages seems not easy either.

File: docs/DeveloperGuide.adoc (hunk 3 above)
Comment (eugenepeh) [reply docs/DeveloperGuide.adoc]: Will end up having 2 sentences start with `see` back to back though,

Is this better?
`Convert the AsciiDoc files in docs to HTML format as instructed in UsingGradle.adoc.`

Comment (Zhiyuan-Amos) [reply docs/DeveloperGuide.adoc]: ok!

File: docs/DeveloperGuide.adoc (hunk 7 above)
Comment (damithc) [reply docs/DeveloperGuide.adoc]: Turns out step 3 is not needed. The fork follows the setting for the upstream repo by default.

Comment (eugenepeh) [reply docs/DeveloperGuide.adoc]: Prof @damithc , you want it to be able to publish gh-pages directly whenever master branch is updated?

Comment (damithc) [reply docs/DeveloperGuide.adoc]: > Prof @damithc , you want it to be able to publish gh-pages directly whenever master branch is updated?

That's how the current repo works.

Comment (damithc) [reply docs/DeveloperGuide.adoc]: Looks like GitHub doesn't publish the gh-pages branch of a fork until we push a new commit to it. 🤔

Comment (eugenepeh) [reply docs/DeveloperGuide.adoc]: oh took me a while to realize that its travis that deploy the gh pages... hmm quite a bit of setup needs to be done,

1. and 2. seems unnecessary since the `gh-pages` branch is there when new users fork the repo,

So perhaps something like this,
1. Go to github settings, choose gh-pages option click save.
2. Go to travis.org, sign in using Github, sync account, enable travis for your organization addressbook level 4.
3. follow the instructions [here (kychua's UsingTravis)](https://github.com/kychua/addressbook-level4/blob/271-migrate-asciidoc/docs/UsingTravis.md) to set up github token to allow travis push to gh-pages branch whenever travis build

Seems like her `UsingTravis` is not in our documentation though > <

Comment (damithc) [reply docs/DeveloperGuide.adoc]: >Go to github settings, choose gh-pages option click save.

Is this necessary? Seems the fork inherits this setting.

>Go to travis.org, sign in using Github, sync account, enable travis for your organization addressbook level 4.

We already have instructions on how to set up Travis. Can refer to that. See section 5.2.

>follow the instructions here (kychua's UsingTravis) to set up github token to allow travis push to gh-pages branch whenever travis build

We need to document this part in UsingTravis.adoc

Comment (pyokagan) [reply docs/DeveloperGuide.adoc]: Yeah, I think it's fine to document the Travis CI automated gh-pages publishing system we have. I think we should document how to set it up *from scratch* (i.e. there is no "upstream" repo), since it would provide a better picture on how things actually work, and safeguards against scenarios such as if future developers need to switch CIs, or Travis CI suddenly forgets our settings or something.

File: docs/DeveloperGuide.adoc [hunk 8]
```diff
@@ -420,7 +420,13 @@ Alternatively, you can download the AsciiDoc plugin for IntelliJ, which allows y
 
 === Publishing Documentation
 
-See <<UsingGithubPages#, UsingGithubPages.adoc>> to learn how to use GitHub Pages to publish documentation to the project site.
+See <<UsingTravis#deploy-github-pages, UsingTravis.adoc>> to learn how to deploy GitHub Pages using Travis.
+
+.  Go to your repository's `Settings -> Options -> GitHub Pages`.
+.  Ensure that the `Source` is set to `gh-pages branch`.
+.  Under `Theme Chooser`, click `Choose a theme` and select a theme of your choice.
```
Comment (damithc) [inline docs/DeveloperGuide.adoc]: Let's ask not to choose themes. That's because choosing a theme creates a commit in the master branch IIRC, diverging their master branch from ours right away. In any case, the theme doesn't matter.

File: docs/DeveloperGuide.adoc [hunk 9]
```diff
@@ -420,7 +420,13 @@ Alternatively, you can download the AsciiDoc plugin for IntelliJ, which allows y
...
+.  Go back to `Settings -> Options -> GitHub Pages`, you should see the message `Your site is ready to be published at \https://<your-username-or-organization-name>.github.io/addressbook-level4/.`
+.  Click the link and you will see your documentation on your GitHub pages.
```
Comment (damithc) [inline docs/DeveloperGuide.adoc]: Did this work for you? If you don't choose a theme, the site will not go live unless you push another commit.

Comment (damithc) [reply docs/DeveloperGuide.adoc]: I think the publishing is triggered by new commit. Either you have to choose a theme (which creates a commit) or push a new commit for the site to go live.

Comment (Zhiyuan-Amos) [inline docs/DeveloperGuide.adoc]: Click **on** the link :)

File: docs/UsingTravis.adoc [hunk 10]
```diff
@@ -54,6 +55,43 @@ image:travis_build.png[Travis build]
 .  If the build is successful, you should be able to check the coverage details of the tests at http://coveralls.io/[Coveralls]
 .  Update the link to the 'build status' badge at the top of the `README.adoc` to point to the build status of your own repo.
 
+== Deploy GitHub Pages
+
+.  Ensure that you have followed the steps above to set up Travis CI.
+.  On GitHub, create a new user account and give this account collaborator access to the repo. Using this account, generate a personal access token https://github.com/settings/tokens/new[here]. *Personal access tokens are like passwords so make sure you keep them secret! If a person access token is leaked, please delete it and generate a new one.*
```
Comment (Zhiyuan-Amos) [inline docs/UsingTravis.adoc]: If a person**al**

File: docs/UsingTravis.adoc [hunk 11]
```diff
@@ -54,6 +55,43 @@ image:travis_build.png[Travis build]
...
+* Add a description for the token (e.g. `Travis CI - deploy docs to gh-pages`)
+* Check the `public_repo` checkbox.
+* Click `Generate Token` and copy your new personal access token.
+--
+We will use this token to grant Travis access to the repo.
+[NOTE]
+We use a new user account to generate the token for team projects to prevent team members from gaining access to other team members' repos. +
+If you are the only one with write access to the repo, you can use your own account to generate the token.
++
+image:generate_token.png[Generate token]
```
Comment (Zhiyuan-Amos) [inline docs/UsingTravis.adoc]: this image should be swapped with the `NOTE`, because the `NOTE` works like an additional note i.e. fyi.

File: docs/DeveloperGuide.adoc (hunk 9 above)
Comment (eugenepeh) [reply docs/DeveloperGuide.adoc]: Tested again, didn't exactly work. After selecting the theme, I will still need to append the url with index.html to see the default page. Only if I push a new commit then the url works without having the index.html.

The index.html doesn't work either unless I chose a theme, which makes it hard for the students to verify whether they've done the correct configuration.

**Comment by eugenepeh** [comment]
can't connect to canihasreview... 
Ready for review

File: docs/DeveloperGuide.adoc [hunk 12]
```diff
@@ -420,7 +420,12 @@ Alternatively, you can download the AsciiDoc plugin for IntelliJ, which allows y
 
 === Publishing Documentation
 
-See <<UsingGithubPages#, UsingGithubPages.adoc>> to learn how to use GitHub Pages to publish documentation to the project site.
+See <<UsingTravis#deploy-github-pages, UsingTravis.adoc>> to learn how to deploy GitHub Pages using Travis. +
+To verify that it works,
+
+.  Update your master branch (e.g. Merging a pull request or etc).
+.  Wait for Travis CI to finish running the build request on your latest commit.
```
Comment (Zhiyuan-Amos) [inline docs/DeveloperGuide.adoc]: finish running the build ~~request~~ right?

File: docs/DeveloperGuide.adoc [hunk 13]
```diff
@@ -420,7 +420,12 @@ Alternatively, you can download the AsciiDoc plugin for IntelliJ, which allows y
...
+.  Go to the URL `\https://<your-username-or-organization-name>.github.io/addressbook-level4/` and you should see your `README` is now the default page.
```
Comment (Zhiyuan-Amos) [inline docs/DeveloperGuide.adoc]: Perhaps can phrase it better:

`Go to the URL \https://<your-username-or-organization-name>.github.io/addressbook-level4/. You should see your README file displayed.`

File: docs/UsingTravis.adoc (hunk 11 above)
Comment (Zhiyuan-Amos) [reply docs/UsingTravis.adoc]: you didn't change this part right? XD

File: docs/UsingTravis.adoc [hunk 14]
```diff
@@ -54,6 +54,43 @@ image:travis_build.png[Travis build]
...
+* Display value in build log: `OFF`
+--
+[NOTE]
+*Make sure you set `Display value in build log` to `OFF`.* +
+Otherwise, other people will be able to see the personal access token and thus have access this repo. +
+Similarly, make sure you *do not print `$GITHUB_TOKEN` to the logs* in Travis scripts as the logs are viewable by the public.
++
+image:travis_add_token.png[Travis add token]
+.  Now, whenever there's a new commit to master branch, Travis will push the latest documentation to gh-pages branch.
+.  If the build is successful, you should be able to see your latest documentation on the gh-pages branch.
```
Comment (Zhiyuan-Amos) [inline docs/UsingTravis.adoc]: what's the difference between this step and the above step?

File: docs/UsingTravis.adoc [hunk 15]
```diff
@@ -54,6 +54,43 @@ image:travis_build.png[Travis build]
...
+* name: `GITHUB_TOKEN`
+* value: personal access token copied in step 1
+* Display value in build log: `OFF`
+--
+[NOTE]
+*Make sure you set `Display value in build log` to `OFF`.* +
+Otherwise, other people will be able to see the personal access token and thus have access this repo. +
+Similarly, make sure you *do not print `$GITHUB_TOKEN` to the logs* in Travis scripts as the logs are viewable by the public.
++
+image:travis_add_token.png[Travis add token]
```
Comment (Zhiyuan-Amos) [inline docs/UsingTravis.adoc]: also, the image here should be swapped with the above `NOTE`? For the same reason mentioned above :>

File: docs/DeveloperGuide.adoc [hunk 16]
```diff
@@ -420,7 +420,12 @@ Alternatively, you can download the AsciiDoc plugin for IntelliJ, which allows y
 
 === Publishing Documentation
 
-See <<UsingGithubPages#, UsingGithubPages.adoc>> to learn how to use GitHub Pages to publish documentation to the project site.
+See <<UsingTravis#deploy-github-pages, UsingTravis.adoc>> to learn how to deploy GitHub Pages using Travis. +
+To verify that it works,
+
+.  Update your master branch (e.g. Merging a pull request or etc).
```
Comment (damithc) [inline docs/DeveloperGuide.adoc]: Suggested:
> To trigger Travis to regenerate documentation, you need to push a new commit to the `master` branch of the fork. Suggested change: Remove the codacy badge from README.adoc

File: docs/UsingTravis.adoc (hunk 15 above)
Comment (eugenepeh) [reply docs/UsingTravis.adoc]: oh 😄 I thought you meant the image inside the NOTE thing

File: docs/DeveloperGuide.adoc (hunk 12 above)
Comment (damithc) [inline docs/DeveloperGuide.adoc]: latest commit -> new commit

File: docs/UsingTravis.adoc (hunk 14 above)
Comment (eugenepeh) [reply docs/UsingTravis.adoc]: oh true... will remove last one

File: docs/DeveloperGuide.adoc [hunk 17]
```diff
@@ -420,7 +420,13 @@ Alternatively, you can download the AsciiDoc plugin for IntelliJ, which allows y
 
 === Publishing Documentation
 
-See <<UsingGithubPages#, UsingGithubPages.adoc>> to learn how to use GitHub Pages to publish documentation to the project site.
+See <<UsingTravis#deploy-github-pages, UsingTravis.adoc>> to learn how to deploy GitHub Pages using Travis. +
+To verify that it works,
+
+.  To trigger Travis to regenerate documentation, you need to push a new commit to the master branch of the fork. +
```
Comment (Zhiyuan-Amos) [inline docs/DeveloperGuide.adoc]: this line doesn't quite fit line 424. :P Maybe something like `Trigger Travis to ... documentation. To do so, you need to...`

File: docs/UsingTravis.adoc [hunk 18]
```diff
@@ -54,6 +54,46 @@ image:travis_build.png[Travis build]
 .  If the build is successful, you should be able to check the coverage details of the tests at http://coveralls.io/[Coveralls]
 .  Update the link to the 'build status' badge at the top of the `README.adoc` to point to the build status of your own repo.
 
+== Deploy GitHub Pages
+
+.  Ensure that you have followed the steps above to set up Travis CI.
+.  On GitHub, create a new user account and give this account collaborator and admin access to the repo. +
+   Using this account, generate a personal access token https://github.com/settings/tokens/new[here]. +
+   *Personal access tokens are like passwords so make sure you keep them secret! If the personal access token is leaked, please delete it and generate a new one.*
```
Comment (Zhiyuan-Amos) [inline docs/UsingTravis.adoc]: This should probably be a `note` instead.

File: docs/UsingTravis.adoc [hunk 19]
```diff
@@ -54,6 +54,48 @@ image:travis_build.png[Travis build]
 .  If the build is successful, you should be able to check the coverage details of the tests at http://coveralls.io/[Coveralls]
 .  Update the link to the 'build status' badge at the top of the `README.adoc` to point to the build status of your own repo.
 
+== Deploy GitHub Pages
```
Comment (damithc) [inline docs/UsingTravis.adoc]: To be consistent with other headings, should be `Deploying`?

**Comment by eugenepeh** [comment]
hmm canihasreview site seems to be up reported by other site but i not able to reach the website 🤕

File: docs/UsingTravis.adoc [hunk 20]
```diff
@@ -54,6 +54,48 @@ image:travis_build.png[Travis build]
 .  If the build is successful, you should be able to check the coverage details of the tests at http://coveralls.io/[Coveralls]
 .  Update the link to the 'build status' badge at the top of the `README.adoc` to point to the build status of your own repo.
 
+== Deploying GitHub Pages
+
+.  Ensure that you have followed the steps above to set up Travis CI.
+.  On GitHub, create a new user account and give this account collaborator and admin access to the repo. +
+   Using this account, generate a personal access token https://github.com/settings/tokens/new[here].
++
+[NOTE]
```
Comment (Zhiyuan-Amos) [inline docs/UsingTravis.adoc]: Does this `NOTE` look weird to you? Cos it's followed by bullet points... XD

Comment (Zhiyuan-Amos) [reply docs/UsingTravis.adoc]: I think we can put it either all the way below (together with the other `NOTE`)? Not sure leh :(

File: docs/UsingTravis.adoc [hunk 21]
```diff
@@ -54,6 +54,48 @@ image:travis_build.png[Travis build]
...
+Personal access tokens are like passwords so make sure you keep them secret! If the personal access token is leaked, please delete it and generate a new one.
++
+--
+* Add a description for the token (e.g. `Travis CI - deploy docs to gh-pages`)
```
Comment (Zhiyuan-Amos) [inline docs/UsingTravis.adoc]: Missing period.

**Comment by Zhiyuan-Amos** [comment]
The rest looks good to me! 👍

File: docs/UsingTravis.adoc (hunk 20 above)
Comment (eugenepeh) [reply docs/UsingTravis.adoc]: hmm yeah I will put together with the other `NOTE` on top :D

**Comment by eugenepeh** [comment]
Ready for review.

**Comment by Zhiyuan-Amos** [comment]
Now I need a merge commit message :P

**Comment by eugenepeh** [comment]
hmm now that's the hard part...
how about

`Update instructions to auto publish GitHub Pages using Travis`?

**Comment by Zhiyuan-Amos** [comment]
Maybe we can change the PR title to `Documentation: Update instructions on deploying to GitHub Pages`, and leave the commit message blank? ^^

**Comment by eugenepeh** [comment]
oo you mean message body? > < I don't think it needs a message body cause this functionality has been there ever since kychua put up the shell code, just that the setting up instruction is not updated.

> Maybe we can change the PR title to Documentation: Update instructions on deploying to GitHub Pages, and leave the commit message blank? ^^

yup sounds good ^ ^
